### ISSUE TITLE
Rust equivalent

### ISSUE THREAD

**Issue by xenide** [billlin0904/xamp2#5]
hey I'm also trying to build a rust equivalent, to do DSD over DoP. But somehow it's not working. Would you be interested to look at some code to see what I'm doing wrong? 

Thanks!

**Comment by billlin0904** [comment]
If you want to convert DSD to DOP for playback:

- On the Windows platform, you must use WASAPI exclusive mode.
- Set the PCM sampling rate (for example, DSD64 requires a sample rate of 176400).
- Encode DSD into the DOP format (for details, please refer to https://dsd-guide.com/dop-open-standard, or use BASS at https://www.un4seen.com/doc/#bass/bassdsd.html).

**Comment by xenide** [comment]
ah you know, I'm using the rust [cpal library](https://github.com/RustAudio/cpal). And in its current state it doesn't support 24 bits, only 32 bits. But DoP requires exactly 24 bits right? I think this may be the reason why it's not working

**Comment by billlin0904** [comment]
You're right, on Windows, DOP playback only supports 24-bit format because it uses WASAPI in exclusive mode. I'm not very familiar with the Rust programming language, but after looking at the implementation in /src/host/wasapi/device.rs, it only implements WASAPI in shared mode, so this implementation cannot achieve DOP playback.
![螢幕擷取畫面 2024-09-08 233610](https://github.com/user-attachments/assets/5c770504-905e-4ce0-9eef-3ee7e920c263)
